services/src/main/java/org/keycloak/authentication/authenticators/client/JWTClientValidator.java

```diff
@@ -0,0 +1,213 @@
+ *
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class JWTClientValidator {
+
+    private static final Logger logger = Logger.getLogger(JWTClientValidator.class);
+
+    private final ClientAuthenticationFlowContext context;
+    private final RealmModel realm;
+    private final int currentTime;
+
+    private MultivaluedMap<String, String> params;
+    private String clientAssertion;
+    private JWSInput jws;
+    private JsonWebToken token;
+    private ClientModel client;
+
+    public JWTClientValidator(ClientAuthenticationFlowContext context) {
+        this.context = context;
+        this.realm = context.getRealm();
+        this.currentTime = Time.currentTime();
+    }
+
+    public boolean clientAssertionParametersValidation() {
+        //KEYCLOAK-19461: Needed for quarkus resteasy implementation throws exception when called with mediaType authentication/json in OpenShiftTokenReviewEndpoint
+        if(!isFormDataRequest(context.getHttpRequest())) {
+            Response challengeResponse = ClientAuthUtil.errorResponse(Response.Status.BAD_REQUEST.getStatusCode(), "invalid_client", "Parameter client_assertion_type is missing");
+            context.challenge(challengeResponse);
+            return false;
+        }
+
+        params = context.getHttpRequest().getDecodedFormParameters();
+
+        String clientAssertionType = params.getFirst(OAuth2Constants.CLIENT_ASSERTION_TYPE);
+        clientAssertion = params.getFirst(OAuth2Constants.CLIENT_ASSERTION);
+
+        if (clientAssertionType == null) {
+            Response challengeResponse = ClientAuthUtil.errorResponse(Response.Status.BAD_REQUEST.getStatusCode(), "invalid_client", "Parameter client_assertion_type is missing");
+            context.challenge(challengeResponse);
+            return false;
+        }
+
+        if (!clientAssertionType.equals(OAuth2Constants.CLIENT_ASSERTION_TYPE_JWT)) {
+            Response challengeResponse = ClientAuthUtil.errorResponse(Response.Status.BAD_REQUEST.getStatusCode(), "invalid_client", "Parameter client_assertion_type has value '"
+                    + clientAssertionType + "' but expected is '" + OAuth2Constants.CLIENT_ASSERTION_TYPE_JWT + "'");
+            context.challenge(challengeResponse);
+            return false;
+        }
+
+        if (clientAssertion == null) {
+            Response challengeResponse = ClientAuthUtil.errorResponse(Response.Status.BAD_REQUEST.getStatusCode(), "invalid_client", "client_assertion parameter missing");
+            context.failure(AuthenticationFlowError.INVALID_CLIENT_CREDENTIALS, challengeResponse);
+            return false;
+        }
+
+        return true;
+    }
+
+    public void readJws() throws JWSInputException {
+        jws = new JWSInput(clientAssertion);
+        token = jws.readJsonContent(JsonWebToken.class);
+    }
+
+    public boolean validateClient() {
+        String clientId = token.getSubject();
```

Comment:
It assumes that `validateClient()` is called after `readJws()` was called. If not, L105 cause NPE.
Mayby the current L105 is OK, but null check might be added.

Reply:
I've added the null check and also added few more null checks on more places. I've used separate commit, so you can easily review changes, but I am planning to squash before merge.

In theory, with correct usage of the JWTClientValidator as is used from the authenticators, this should not happen. But it is indeed good to add null checks (for instance for the case when someone wants to override default authenticators and use the validators in his code etc).

Reply:
Thank you.